### ISSUE TITLE
Modification of non-creatable hash value attempted

### ISSUE THREAD

**Issue by tsibley** [rurban/Cpanel-JSON-XS#45]
Cpanel::JSON::XS has started failing this simple test case:

```perl
use strict;
use warnings;
use Test::More;
use Cpanel::JSON::XS;

my $json = '{"foo":null}';
my $data = Cpanel::JSON::XS->new->decode($json);

ok exists $data->{foo}, "foo exists";
is $data->{foo}, undef, "foo is undef";
ok $data->{foo} = "bar", "foo can be set to 'bar'";

done_testing;
```

during the third `ok` with the Perl error `Modification of non-creatable hash value attempted`.

3.0104 didn't exhibit this error, so I bisected the failure to 752377a ("bool: return the same true and false objects"):

```
$ git bisect start master 3.0104
…
$ git bisect run bash -c 'perl Makefile.PL && make && prove --blib -v json-null.t'
…
json-null.t .. 
ok 1 - foo exists
ok 2 - foo is undef
Modification of non-creatable hash value attempted, subscript "foo" at json-null.t line 11.
# Tests were run but no plan was declared and done_testing() was not seen.
# Looks like your test exited with 255 just after 2.
Dubious, test returned 255 (wstat 65280, 0xff00)
All 2 subtests passed 

Test Summary Report
-------------------
json-null.t (Wstat: 65280 Tests: 2 Failed: 0)
  Non-zero exit status: 255
  Parse errors: No plan found in TAP output
Files=1, Tests=2,  0 wallclock secs ( 0.02 usr  0.02 sys +  0.02 cusr  0.00 csys =  0.06 CPU)
Result: FAIL
752377a893e0b80566a752826719e4025972ebae is the first bad commit
commit 752377a893e0b80566a752826719e4025972ebae
Author: Reini Urban <rurban@cpanel.net>
Date:   Wed Nov 25 12:39:50 2015 +0100

    bool: return the same true and false objects
    
    not a fresh sv was was previosulty a ref to true/false

:100644 100644 1ed68413802ec75f791558c1a518b9b56bee99b3 7e2a061ab9f887d2fcb468461e57616365ae77f9 M	XS.xs
bisect run success
```

The changes in the commit make it clear that `true` and `false` values exhibit the same issue as `null`, which you can modify the test file above to see.

**Comment by tsibley** [comment]
A version of the test file which also exercises `true/false`:

```perl
use strict;
use warnings;
use Test::More;
use Cpanel::JSON::XS;

eval {
    my $json = '{"foo":null}';
    my $data = Cpanel::JSON::XS->new->decode($json);

    ok exists $data->{foo}, "foo exists";
    is $data->{foo}, undef, "foo is undef";
    ok $data->{foo} = "bar", "foo can be set to 'bar'";
};
ok !$@, "No error" or diag "Error: $@";

eval {
    my $json = '{"foo":true}';
    my $data = Cpanel::JSON::XS->new->decode($json);

    ok exists $data->{foo}, "foo exists";
    ok $data->{foo}, "foo is true";
    ok $data->{foo} = "bar", "foo can be set to 'bar'";
};
ok !$@, "No error" or diag "Error: $@";

eval {
    my $json = '{"foo":false}';
    my $data = Cpanel::JSON::XS->new->decode($json);

    ok exists $data->{foo}, "foo exists";
    ok !$data->{foo}, "foo is false";
    ok $data->{foo} = "bar", "foo can be set to 'bar'";
};
ok !$@, "No error" or diag "Error: $@";

done_testing;
```

**Comment by karenetheridge** [comment]
See also https://rt.cpan.org/Ticket/Display.html?id=109899

**Comment by rurban** [comment]
Oh my, sorry. I thought returning uncopied values of undef, true and false is ok. We really need a copy.
Fixed in `mod-hvalue-gh45`

**Comment by rurban** [comment]
Merged to master

**Comment by pghmcfc** [comment]
The new test t/55_modifiable.t is failing for me with Perl < 5.12:
```
$ make test AUTHOR_TESTING=1
PERL_DL_NONLAZY=1 /usr/bin/perl "-MExtUtils::Command::MM" "-e" "test_harness(0, 'blib/lib', 'blib/arch')" -Mcommon::sense t/*.t
t/00_load.t ................. ok
t/01_utf8.t ................. ok
t/02_error.t ................ ok
t/03_types.t ................ ok
t/04_dwiw_encode.t .......... ok
t/05_dwiw_decode.t .......... ok
t/06_pc_pretty.t ............ ok
t/07_pc_esc.t ............... ok
t/08_pc_base.t .............. ok
t/08_pc_base_nv.t ........... ok
t/09_pc_extra_number.t ...... ok
t/10_pc_keysort.t ........... ok
t/117_numbers.t ............. ok
t/11_pc_expo.t .............. ok
t/12_blessed.t .............. ok
t/13_limit.t ................ ok
t/14_latin1.t ............... ok
t/15_prefix.t ............... ok
t/16_tied.t ................. ok
t/17_relaxed.t .............. ok
t/18_json_checker.t ......... ok
t/19_incr.t ................. ok
t/20_faihu.t ................ ok
t/20_unknown.t .............. ok
t/21_evans.t ................ ok
t/22_comment_at_eof.t ....... ok
t/23_array_ctx.t ............ ok
t/24_freeze_recursion.t ..... ok
t/25_boolean.t .............. ok
t/52_object.t ............... ok
t/53_readonly.t ............. ok
t/54_stringify.t ............ ok
#   Failed test 'foo hvalue true'
#   at t/55_modifiable.t line 16.
#          got: 'true'
#     expected: '1'
# Looks like you failed 1 test of 15.
t/55_modifiable.t ........... 
Dubious, test returned 1 (wstat 256, 0x100)
Failed 1/15 subtests 
t/96_interop.t .............. ok
t/96_mojo.t ................. skipped: Mojo::JSON required for testing interop
t/97_unshare_hek.t .......... ok
t/98_56only.t ............... ok
t/99_binary.t ............... ok
t/z_kwalitee.t .............. ok
t/z_leaktrace.t ............. ok
t/z_meta.t .................. ok
t/z_perl_minimum_version.t .. ok
t/z_pod-coverage.t .......... ok
t/z_pod.t ................... ok
Test Summary Report
-------------------
t/55_modifiable.t         (Wstat: 256 Tests: 15 Failed: 1)
  Failed test:  5
  Non-zero exit status: 1
Files=44, Tests=31947,  8 wallclock secs ( 1.76 usr  0.10 sys +  7.42 cusr  0.17 csys =  9.45 CPU)
Result: FAIL
Failed 1/44 test programs. 1/31947 subtests failed.
make: *** [test_dynamic] Error 255
```
Seen this with 5.8.8, 5.10.0 and 5.10.1. All my tests on 5.12.4 and later pass.

**Comment by rurban** [comment]
Interesting, cannot repro this behavior. But it's no big deal, I'll relax this test to accept both values.
Thanks.

**Comment by pghmcfc** [comment]
Hmm, might be something Fedora/Red Hat specific involved (seen with Fedora < 14 and CentOS < 7).

Or possibly related to an old version of some module that's being pulled in.

On an unrelated note, JSON.pm is now a required test dependency in t/20_unknown.t, whilst its use in t/96_interop.t is optional. Not sure if that was intended.

**Comment by rurban** [comment]
Fixed with a6af4d85302d44d145fed0702eb50dada07e6ee0 and 96b2fec7983bf6a2f806b92c2f4abe4c7290397b, thanks. Smoking at https://travis-ci.org/rurban/Cpanel-JSON-XS/builds/93960684
